popovers: Added search typeahead to Move topic popover.

aryanshridhar opened the pull request:
Added an input typeahead within the `Move Topic` modal, for the user to select a desired stream to move a specific topic.
The purpose it to reduce users effort since a dropdown makes it time consuming to search for a stream (especially when dealing with a plenty of streams).

<strong>Screenshots :</strong>
![movefix](https://user-images.githubusercontent.com/53977614/108750109-a75dfa00-7566-11eb-9f3c-c0022f1ab965.gif)


Closes #14860.

Would appreciate a review @amanagr.

amanagr commented:
2 things:
* We don't want the dropdown to select streams now.
* We don't need the select button. Just enter / click to automatically select.

aryanshridhar commented:
Hey @amanagr, Should the state of dropdown be set to <code>readonly</code>, since removing the entire dropdown UI would also detach the feature of changing topic names.

amanagr commented:
We want to replace the dropdown with the stream typeahead input field. Is that somehow not possible? (We don't want the select stream button too.)

aero31aero commented:
Also, we should reuse `DropdownListWidget` for this. That was created for usecases like this one.

aryanshridhar commented:
Would be great if @amanagr and @aero31aero could have a look at it again.(The screenshots are updated in the pr description).

amanagr commented:
Can you make `enter` to select work? Also, you should exclude the currently selected stream from the stream_list.

aryanshridhar commented:
Hey @amanagr, `enter` does seem to work well for selecting stream, but fails to update the `stream_header_colorblock`.(Fixed that, thanks!)
However, In order to exclude the currently selected stream, I need to explicitly remove the stream from the dropdown (maybe by `$('#element').remove()`). It is so because, `DropdownListWidget` tends to find an <a href= 'https://github.com/zulip/zulip/blob/master/static/js/dropdown_list_widget.js#L36'>`item`</a> which takes up the whole array of available objects as a parameter (here, `stream_list`) and renders it. A similar behaviour is observed while editing bots which also extends the  `DropdownListWidget`. (<a href = 'https://user-images.githubusercontent.com/53977614/108872224-f451d700-761f-11eb-9557-8f6b092639cf.gif'>Attached GIF</a>).

Hence, Should I adopt the current method (explicitly writing `jquery` to remove it from DOM) or update the current behaviour of `DropdownListWidget`(by creating a separate pr, or adding a commit to this one) ?

timabbott commented on static/js/dropdown_list_widget.js:
```diff
@@ -77,8 +77,14 @@ const DropdownListWidget = function (opts) {
+                return opts.data;
+            }
+            return opts.data.filter((x) => x.value !== opts.value.toString());
+        };
```
Will this have an effect on other places we use this widget?  @aero31aero thoughts would be appreciated for whether this is a good idea or something that should be controlled as an option.

aryanshridhar replied:
Yep, it will affect other places as well.
If the fact of excluding the current list item from the dropdown doesn't seem right, is it better to have an extra object (parameter) while passing it to `dropdown_list_widget`.

Something like: 
(Will exclude the current list item from the dropdown.)
```javascript
const opts = {
  retain_current_item: false,
};

dropdown_list_widget(opts);
```

having a default value of `true`.

timabbott replied:
Yeah, let's do that parameter for now - that way we can contain the complexity of this change, and then please start a thread on #frontend about whether we want to change the value of that parameter for other users of this widget.

aryanshridhar replied:
Cool! Will refactor it.

timabbott commented on static/styles/app_components.css:
```diff
@@ -1,7 +1,7 @@
    (not included in the portico CSS) */
 
-.flex {
+   .flex {
```
This looks like a typo.

timabbott commented on static/templates/settings/dropdown_list_widget.hbs:
```diff
@@ -17,7 +17,9 @@
     <button class="button small dropdown_list_reset_button">
         <a>{{reset_button_text}}</a>
     </button>
+    {{/if}}
```
This should probably be pulled out as a separate preparatory fix for `dropdown_list_widget`, I think.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -367,6 +372,15 @@ exports.register_click_handlers = function () {
+            Number.parseInt(stream_widget.value(), 10),
+        );
+        ui_util.decorate_stream_bar(stream_name, stream_header_colorblock, false);
+    });
```
Can you add a comment on why this doesn't use `stopPropagation` list most of our click handlers?

timabbott commented on static/styles/app_components.css:
```diff
@@ -1,7 +1,7 @@
    (not included in the portico CSS) */
 
-   .flex {
+.flex {
```
Ahh, this fixup is just in the wrong commit -- it's important to read your commits with a tool like `gitk` and make sure you put fixes into the right commit.

timabbott commented on static/js/stream_popover.js:
```diff
@@ -372,9 +372,12 @@ exports.register_click_handlers = function () {
 
+        if (e.type === "keypress" && e.which !== 13) {
+            return;
+        }
```
This change seems to be not related to the "overflow" description.

aryanshridhar replied:
Ahh, Apologies for these silly commit errors!

timabbott commented:
@aryanshridhar I think you forgot to push your changes -- I don't see any updates after my last batch of comments.

aryanshridhar commented:
Thanks @timabbott, Updated the pr with the following changes: 
<ul>
<li>Commit 0befec9bae620771c204892726a912b4dc19da00: Wrapped the entire <code>reset_button_group</code> that renders it only if <code>reset_button_text</code> is passed to the template.</li>
<li>Commit 117757014c21bd69fe999af6606a1285b3b7cc5e: Added the <code>retain_current_item</code> parameter to <code>dropdown_list_widget</code> as it was discussed <a href = 'https://github.com/zulip/zulip/pull/17216#discussion_r597491717'>here</a>. (Will open up a thread in #frontend to gather more information about it)
<li>Commit f07c643a893711797c3814c7e494186f6e86abdf: Adds the <code>dropdown_list_widget</code> to <code>Move-topic</code> popovers, fixing the issue #14860.
<li>Commit 6535660d1ce989c9b416c31552c5a16e473f2195: A small followup of f07c643a893711797c3814c7e494186f6e86abdf, to make the <code>dropdown_list_widget</code> overflow the modal. 
</ul>

Apologies for the previously wrong commits. On the brighter side, I got to learn about <code>gitk</code> and its uses. Thanks!

amanagr commented on static/styles/settings.css:
```diff
@@ -632,11 +632,20 @@ input[type="checkbox"] {
 }
```
We also need 
```
border-left: 0px;
border-radius: 0px 4px 4px 0px;

```
for the stream box so that it looks like this:
<img width="375" alt="Screenshot 2021-03-30 at 3 04 02 PM" src="https://user-images.githubusercontent.com/25124304/112967712-2d262400-9169-11eb-830f-a63dd6defe89.png">

amanagr reviewed (commented):
@aryanshridhar thanks for updating. Can you make so that after clicking on the stream box, the focus should be set on search automatically?

amanagr commented:
I found that moving messages between stream for topics is broken.

Steps to reproduce:
1. Narrow to a topic
2. Try to move a topic which you are not narrowed to, it will run into some error due to message being undefined. I am working on a fix.

aryanshridhar commented:
> @aryanshridhar thanks for updating. Can you make so that after clicking on the stream box, the focus should be set on search automatically?

Thanks, However it seems like I need to add a commit specifically for `dropdown_list_widget` to make the focus thing work, since it isn't the case for other places where  `dropdown_list_widget` are implemented and feels like a nice improvement as well.

amanagr commented:
Yeah, go ahead!

amanagr commented:
Opened #17888

timabbott commented:
I merged #17888; let me know when this is rebased on that and ready.

aryanshridhar commented:
Rebased the pr @timabbott. 
Following the changes <a href = 'https://github.com/zulip/zulip/pull/17216#issuecomment-808938647'>previously</a> made, here are the ones that are added --

<ul>

<li>Commit 8e80ba6f34a95396fdf225f8f2d6e4d1a255d2ef: Added an ability to set focus on input field when the user toggles the <code>dropdown_list_widget</code> button.</li>
<li>Commit 2193eb1ca39b701b4211f813ec353800360350b6: Enhanced style of <code>Move topic</code> dropdown button to look similar to what @amanagr suggested <a href = 'https://github.com/zulip/zulip/pull/17216#discussion_r603938146'>here</a></li>
</ul>

timabbott commented:
I'll wait for a rebase and another look by @amanagr before doing another detailed review on this one.  (I just merged your other `dropdown_list_widget` PR, which caused the merge conflict).

aryanshridhar commented:
Thanks, Rebased it :)

timabbott commented:
I merged the first commit.  Will post a few comments.

amanagr commented on static/js/dropdown_list_widget.js:
```diff
@@ -119,7 +120,20 @@ export const DropdownListWidget = function ({
-            search_input.trigger("focus");
+            // We need to trigger the focus event once the input field is
+            // properly bind to the DOM.
+            timeout = setTimeout(() => {
```
> Does bootstrap or can we trigger some event when the modal is displayed and then set focus? This doesn't seem like the right way to go for this.

aryanshridhar replied:
Updated the changes as discussed within 3ff10553dba95b64f40b2c3ec5b7d245cd13860a.

aryanshridhar commented on static/js/dropdown_list_widget.js:
```diff
@@ -8,6 +8,7 @@ import * as ListWidget from "./list_widget";
 export const DropdownListWidget = function ({
     widget_name,
     data,
+    retain_current_item = true,
```
Yea, `include_current_item` does sound better.
What about `exclude_current_item`? Since adding a default parameter of `false` and later setting that to `true` within it's use case feels more natural. 
However, kept `include_current_item` for now. As discussed earlier, Will open up a thread as a follow-up work regarding the same.

aryanshridhar commented:
Rebased it again to pass the Codeql errors :)

timabbott commented on static/styles/app_components.css:
```diff
@@ -407,7 +407,7 @@
 .stream_header_colorblock {
     @extend .stream-selection-header-colorblock;
-    margin-bottom: 5px;
+    margin-bottom: 20px;
```
I'm pretty sure this is going to break any other widget using the `stream_header_colorblock` CSS class.

aryanshridhar replied:
Thanks, I should have checked this obvious change within other places as well.

aryanshridhar replied:
Fixed this in c343b1c98b6660eb3a06bcb1afd24c2741edb1f4.

timabbott commented:
I merged the first commit as c630b888e508348f2bf644307f41ffb7a5ad05e0, since that part is clearly useful.  I don't think I understand the rest of this PR -- for this:
```
    dropdown_list_widget: Properly bind focus event to input field.
    
    Earlier, the focus event was triggering on a hidden
    dropdown input field, causing it to remain blur.
    Hence, Added a fix by explicitly triggering the focus event
    when the input field is visible in DOM.
```

Can you explain clearly what the bug was?  What does "remain blur" mean here?

timabbott commented on static/templates/move_topic_to_stream.hbs:
```diff
@@ -1,9 +1,9 @@
         <h3 id="move_topic_modal_label">{{t "Move topic" }} </h3>
     </div>
-    <div class="modal-body">
+    <div class="modal-body move-topic-modal">
```
Why do you need to add this class in two places?

aryanshridhar replied:
This is so because `modal-body` and `modal` classes doesn't support the overlay objects to overflow.Hence, had to explicitly mention the `modal-body` to overflow here.
Here's the GIF if an extra `overflow` CSS property isn't attached to it. 

![overflow](https://user-images.githubusercontent.com/53977614/114082549-83166c80-98cb-11eb-9222-66192c62399e.gif).

aryanshridhar commented:
> Can you explain clearly what the bug was? What does "remain blur" mean here?

<strong>Issue</strong>
Previously, the focus event wasn't triggering on  on the dropdown input field when the `dropdown_button` was clicked.

Here's an GIF replicating the bug: (Since, pictures are worth a thousand words!)
![old](https://user-images.githubusercontent.com/53977614/113917462-dcfd3080-97fe-11eb-85e6-2995fac6b483.gif)
(Notice, the focus should have been automatically triggered to input field)

The responsible code for this was: 
```javascript
dropdown_toggle.on("focus", (e) => {
    // On opening a Bootstrap Dropdown, the parent element receives focus.
    // Here, we want our search input to have focus instead.
    e.preventDefault();
    search_input.trigger("focus");
});
```
<strong>Solution</strong>

The issue here was that when the click handler was initially triggered on `focus` event, the input field was not visible in the DOM.
Due to this issue, the input field wasn't gaining focus to it (Since, focus can only be triggered to elements that are visible in the DOM).

Thus, the fix here was by adding a check statement to trigger the focus on input field <strong>only</strong> if it's  visible in the DOM. 
The code now looks like: 

```javascript
  dropdown_toggle.on("focus", (e) => {
      // On opening a Bootstrap Dropdown, the parent element receives focus.
      // Here, we want our search input to have focus instead.
      e.preventDefault();
      // Trigger the focus event once the input field is visible.
      if (dropdown_list_body.is(":visible")) {
          search_input.trigger("focus");
      }
  });
```

The GIF now looks like: 

![fix-focus](https://user-images.githubusercontent.com/53977614/113919082-cd7ee700-9800-11eb-9c97-9e5b1767d2c9.gif)

As it can be seen the focus event is properly binded with the input field and it is is gaining focus as it should have been.

aryanshridhar commented:
Rebased the pr.
Also, added a new commit d3772c6dd4ef2574cd40f7c196f980e65323a54b as a follow up of #17926.

timabbott commented on static/styles/zulip.css:
```diff
@@ -2357,6 +2357,12 @@ div.topic_edit_spinner .loading_indicator_spinner {
     max-height: 60vh;
 }
 
+.move-topic-modal {
```
Just use `#move_topic_modal` for this.

timabbott commented on static/templates/move_topic_to_stream.hbs:
```diff
@@ -1,4 +1,4 @@
-<div id="move_topic_modal" class="modal modal-bg hide fade new-style" tabindex="-1" role="dialog" aria-labelledby="move_topic_modal_label" aria-hidden="true">
+<div id="move_topic_modal" class="modal move-topic-modal modal-bg hide fade new-style" tabindex="-1" role="dialog" aria-labelledby="move_topic_modal_label" aria-hidden="true">
```
You don't need to add this duplicate class here.

timabbott commented:
I merged this, after adding a few cleanup commits and fixing the use of a `move_topic_popover` CSS class for two unrelated elements that you wanted the same property on, which made the code very confusing, and several other minor cleanups to better scope selectors.

There's a natural follow-up project of using the same `dropdown_list_widget` for the stream selector in the main message-edit UI; can you do that as a follow-up?

timabbott commented:
I opened https://github.com/zulip/zulip/issues/18415 for a different followup I noticed while reviewing this.

timabbott commented:
And opened https://github.com/zulip/zulip/issues/18416 for the dropdown_list_widget work, in case you don't end up doing that one.

aryanshridhar commented:
Cool! Will tackle the latter one as of now.
